[CheckpointWrapper] Decouple CPU offload

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85062
* #84908
* __->__ #84907

This fixes the activation offload for checkpoint wrapper, which was previously broken. It was broken because it was tightly coupled with activation checkpoint, i.e. we did:

```
with save_on_cpu:
    checkpoint(module_forward())
```

which would not offload any activation tensors to CPU, as those activations would already be not saved by autograd due to the checkpoint implementation taking priority. 

Now, if `offload_to_cpu` is specified, we only do `save_on_cpu` and no checkpoint, so all intermediate tensors are offloaded to CPU instead of checkpointed. 

These wrappers can be composed, i.e. if we have

`(Linear, Linear) -> (Linear, Linear) -> (Linear, Linear)`

we can do

`Offload( checkpoint(Linear, Linear) -> checkpoint(Linear, Linear) -> checkpoint(Linear, Linear))`

and inner tensors would be checkpointed while outers will be offloaded.

Differential Revision: [D39448882](https://our.internmc.facebook.com/intern/diff/D39448882/)

rohan-varma committed:
[CheckpointWrapper] Decouple CPU offload



Differential Revision: [D39448882](https://our.internmc.facebook.com/intern/diff/D39448882/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[CheckpointWrapper] Decouple CPU offload"


Differential Revision: [D39448882](https://our.internmc.facebook.com/intern/diff/D39448882/)

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_checkpoint_wrapper.py:
```diff
@@ -189,5 +189,47 @@ def test_fqn(self):
+                    self.assertEqual(torch.device("cpu"), saved.device)
+
+            if hasattr(grad_fn, 'next_functions'):
+                for k, _ in grad_fn.next_functions:
```
nit: Does `k` stand for something? Maybe we can use a more descriptive name here since the 1st element of each tuple in `next_functions` is always a backward function (I think)?

awgu commented on test/distributed/fsdp/test_checkpoint_wrapper.py:
```diff
@@ -189,5 +189,47 @@ def test_fqn(self):
+
+                saved = getattr(grad_fn, e)
+                if isinstance(saved, torch.Tensor):
+                    self.assertEqual(torch.device("cpu"), saved.device)
```
Do you think it is worthwhile to check that we run this `assertEqual()` at least once or some expected number of times (to make sure the DFS is achieving what we expect)?

rohan-varma replied:
Added a check to ensure we call it at least once.

awgu commented on torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:
```diff
@@ -18,7 +17,9 @@ class CheckpointImpl(Enum):
-    An nn.Module that wraps another nn.Module with checkpointing.
+    An nn.Module that wraps another nn.Module with checkpointing. Note that this
+    module is not meant to be used directly, but instead it is to be used
+    through the ``checkpoint_wrapper`` function.
```
I guess it is too late to make this private `_CheckpointWrapper`.

rohan-varma replied:
I think we can discuss this further and do in follow up work, as this API is still private / prototype.

awgu commented on torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:
```diff
@@ -148,8 +154,10 @@ def checkpoint_wrapper(
+            Whether to offload activations of this wrapped module to CPU. Note
+            that if this is specified, ``checkpoint_impl`` and ``checkpoint_fn``
+            arguments will be ignored in favor of the activations being
+            offloaded to CPU. Default is ``False``.
```
I have not been too in the loop. I am wondering why we are coupling these  `offload_to_cpu=True` and `offload_to_cpu=False` options under the same `checkpoint_wrapper()`. Why not an `activation_offload_wrapper()` and `activation_checkpoint_wrapper()` or something?

rohan-varma replied:
I think we can consider the latter. This API is still in prototype stage and I wanted something that's easy to flip back and forth, but definitely agree usability wise the current options aren't what we want for a stable end product, it is too confusing with certain args only working with others and otherwise being ignored

rohan-varma committed:
Update on "[CheckpointWrapper] Decouple CPU offload"


This fixes the activation offload for checkpoint wrapper, which was previously broken. It was broken because it was tightly coupled with activation checkpoint, i.e. we did:

```
with save_on_cpu:
    checkpoint(module_forward())
```

which would not offload any activation tensors to CPU, as those activations would already be not saved by autograd due to the checkpoint implementation taking priority. 

Now, if `offload_to_cpu` is specified, we only do `save_on_cpu` and no checkpoint, so all intermediate tensors are offloaded to CPU instead of checkpointed. 

These wrappers can be composed, i.e. if we have

`(Linear, Linear) -> (Linear, Linear) -> (Linear, Linear)`

we can do

`Offload( checkpoint(Linear, Linear) -> checkpoint(Linear, Linear) -> checkpoint(Linear, Linear))`

and inner tensors would be checkpointed while outers will be offloaded.

Differential Revision: [D39448882](https://our.internmc.facebook.com/intern/diff/D39448882/)

[ghstack-poisoned]

awgu reviewed (commented):
Leaving some initial comments

rohan-varma committed:
Update on "[CheckpointWrapper] Decouple CPU offload"


This fixes the activation offload for checkpoint wrapper, which was previously broken. It was broken because it was tightly coupled with activation checkpoint, i.e. we did:

```
with save_on_cpu:
    checkpoint(module_forward())
```

which would not offload any activation tensors to CPU, as those activations would already be not saved by autograd due to the checkpoint implementation taking priority. 

Now, if `offload_to_cpu` is specified, we only do `save_on_cpu` and no checkpoint, so all intermediate tensors are offloaded to CPU instead of checkpointed. 

These wrappers can be composed, i.e. if we have

`(Linear, Linear) -> (Linear, Linear) -> (Linear, Linear)`

we can do

`Offload( checkpoint(Linear, Linear) -> checkpoint(Linear, Linear) -> checkpoint(Linear, Linear))`

and inner tensors would be checkpointed while outers will be offloaded.

Differential Revision: [D39448882](https://our.internmc.facebook.com/intern/diff/D39448882/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[CheckpointWrapper] Decouple CPU offload"


This fixes the activation offload for checkpoint wrapper, which was previously broken. It was broken because it was tightly coupled with activation checkpoint, i.e. we did:

```
with save_on_cpu:
    checkpoint(module_forward())
```

which would not offload any activation tensors to CPU, as those activations would already be not saved by autograd due to the checkpoint implementation taking priority. 

Now, if `offload_to_cpu` is specified, we only do `save_on_cpu` and no checkpoint, so all intermediate tensors are offloaded to CPU instead of checkpointed. 

These wrappers can be composed, i.e. if we have

`(Linear, Linear) -> (Linear, Linear) -> (Linear, Linear)`

we can do

`Offload( checkpoint(Linear, Linear) -> checkpoint(Linear, Linear) -> checkpoint(Linear, Linear))`

and inner tensors would be checkpointed while outers will be offloaded.

Differential Revision: [D39448882](https://our.internmc.facebook.com/intern/diff/D39448882/)

[ghstack-poisoned]

awgu reviewed (approved):
LGTM!

rohan-varma committed:
Update on "[CheckpointWrapper] Decouple CPU offload"


This fixes the activation offload for checkpoint wrapper, which was previously broken. It was broken because it was tightly coupled with activation checkpoint, i.e. we did:

```
with save_on_cpu:
    checkpoint(module_forward())
```

which would not offload any activation tensors to CPU, as those activations would already be not saved by autograd due to the checkpoint implementation taking priority. 

Now, if `offload_to_cpu` is specified, we only do `save_on_cpu` and no checkpoint, so all intermediate tensors are offloaded to CPU instead of checkpointed. 

These wrappers can be composed, i.e. if we have

`(Linear, Linear) -> (Linear, Linear) -> (Linear, Linear)`

we can do

`Offload( checkpoint(Linear, Linear) -> checkpoint(Linear, Linear) -> checkpoint(Linear, Linear))`

and inner tensors would be checkpointed while outers will be offloaded.

Differential Revision: [D39448882](https://our.internmc.facebook.com/intern/diff/D39448882/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[CheckpointWrapper] Decouple CPU offload"


This fixes the activation offload for checkpoint wrapper, which was previously broken. It was broken because it was tightly coupled with activation checkpoint, i.e. we did:

```
with save_on_cpu:
    checkpoint(module_forward())
```

which would not offload any activation tensors to CPU, as those activations would already be not saved by autograd due to the checkpoint implementation taking priority. 

Now, if `offload_to_cpu` is specified, we only do `save_on_cpu` and no checkpoint, so all intermediate tensors are offloaded to CPU instead of checkpointed. 

These wrappers can be composed, i.e. if we have

`(Linear, Linear) -> (Linear, Linear) -> (Linear, Linear)`

we can do

`Offload( checkpoint(Linear, Linear) -> checkpoint(Linear, Linear) -> checkpoint(Linear, Linear))`

and inner tensors would be checkpointed while outers will be offloaded.

Differential Revision: [D39448882](https://our.internmc.facebook.com/intern/diff/D39448882/)

[ghstack-poisoned]
